### PR TITLE
[SDK-1077]: add request when available in enableLogging

### PR DESCRIPTION
**Pull request by code-crusher**
## Reference
SDK-1077: [android] Update native SDK to report X-Branch-Request-ID when present

## Description
Print request-id in the logs when;
1. EnableLogging is set
2. X-Branch-Request-Id-Header is present in the response header

```
I/BranchSDK: returned request-id: [ 1231231__!32123123 ] with {"session_id":"849308671259384261","identity_id":"849308509983354059","link":"https://bnctestbed.test-app.link?%24identity_id=849308509983354059","data":"{\"+clicked_branch_link\":false,\"+is_first_session\":false}","device_fingerprint_id":"849308509959273715"}
```

## Testing Instructions
- [X] Manual

## Risk Assessment [`LOW`]

- [X] I, the PR creator, have tested — integration, unit, or otherwise — this code.

## Reviewer Checklist (To be checked off by the reviewer only)

- [ ] JIRA Ticket is referenced in PR title.
- Correctness & Style
    - [ ] Conforms to [AOSP Style Guides](https://source.android.com/setup/contribute/code-style)
    - [ ] Mission critical pieces are documented in code and out of code as needed.
- [ ] Unit Tests reviewed and test issue sufficiently.
- [ ] Functionality was reviewed in QA independently by another engineer on the team.

cc @BranchMetrics/core-team

### CODE REVIEW

**Comment by jdee** [comment]
So I started to do this sort of thing for iOS, adding a `requestId` field to a request object somewhere. But I ended up going with a two-line change that just dumps all response headers to the log. For the Web SDK, we'll probably make no change, since you can just see the header in the built-in browser tools. I wonder what the best approach is here. @bklastaitis-branch any opinions?

**Comment by code-crusher** [comment]
Just dumping all the headers is not really UX friendly(and also we only need a single header to be printed), the above format will clearly specify what is the request ID connected with the response.

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java
Comment (bklastaitis-branch) [inline Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java:120]: Variable declaration can be pulled into the try block (where value is assigned) for brevity.

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java
Comment (bklastaitis-branch) [inline Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java:220]: We don't need this method, it can be safely replaced with a oneliner: `connection.getHeaderField("X-Branch-Request-Id");` which does the same thing as the method. However, the string `"X-Branch-Request-Id"` should be pulled out and defined as one of the values in a new enum `Headers` in the `Defines` class.

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java
Comment (bklastaitis-branch) [inline Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java:52]: Same as below: Variable declaration can be pulled into the try block (where value is assigned) for brevity.

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java
Comment (bklastaitis-branch) [inline Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java:182]: Can we replace all arguments in `processEntityForJSON` with just `BranchResponse` and `tag`? This way it's clear that `processEntityForJSON` converts from the `BranchResponse` abstraction to json-formatted `ServerResponse`. Also, for brevity, let's get rid of `serverResponse.setRequestId_` and assign that value in the constructor, so it becomes a read-only variable, just like `ServerResponse.statusCode_`. In order to avoid empty strings we can use `Strings.emptyToNull()` when passing the request id to to the `ServerResponse` constructor

File: Branch-SDK/src/main/java/io/branch/referral/ServerResponse.java
Comment (bklastaitis-branch) [inline Branch-SDK/src/main/java/io/branch/referral/ServerResponse.java:120]: Let's get rid of this, see the comment on `processEntityForJSON` for more info

File: .circleci/config.yml
Comment (bklastaitis-branch) [inline .circleci/config.yml:10]: I assume changes in this file were needed but I'm not really following what they. Would you describe why you added them? PR description/Jira ticket didn't mention anything.

**Review by bklastaitis-branch — CHANGES_REQUESTED** [review]
@code-crusher I left some comments, requested mainly stylistic changes but I think they are important. It's not necessarily related to your code but the SDK is full of little design quirks like that, which make it harder to read and understand the code.

**Comment by bklastaitis-branch** [comment]
@jdee I'd lean towards @code-crusher 's approach. UX wise, the reader of the logs is most likely only interested in the request id, not any other headers, so printing other headers just makes it harder to read/find the request id.

As for creating the request id entity in the first plcase - rather than just printing and forgetting - is fair game I think. We may one day use that request id, especially for things like fraud detection.

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java
Comment (code-crusher) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java:120]: Fair

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java
Comment (code-crusher) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java:182]: 1. Yeah, let's do Object here, did not wanted change much but will update.
2. Already using `!TextUtils.isEmpty(requestId)` -- this checks for both null and empty. Not sure how using `Strings.emptyToNull()` will help since we have to check for null in any case, the first method does both.

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java
Comment (code-crusher) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java:220]: Yeah, missed the constant, will update

File: .circleci/config.yml
Comment (code-crusher) [reply .circleci/config.yml:10]: The CI was failing, we were using an old and incorrect config. This is just basic CI config for Android. Cache will make CI builds faster since the same dependecies will not be pulled again and again.

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java
Comment (code-crusher) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java:52]: I remember this now, this is being used outside of try-catch also, that's why put it here.

File: Branch-SDK/src/main/java/io/branch/referral/ServerResponse.java
Comment (code-crusher) [reply Branch-SDK/src/main/java/io/branch/referral/ServerResponse.java:120]: I created this to keep usages in the future if someone want to pull thisi id apart from the logs, but I'm fine to remove it

**Comment by code-crusher** [comment]
@bklastaitis-branch updated, if this looks good, can we merge?

**Comment by ahmednawar** [comment]
Please check that Android and iOS are consistent in how they show this ID to developers. cc @jdee

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java
Comment (bklastaitis-branch) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java]: My bad didn't notice it

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java
Comment (bklastaitis-branch) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java:219]: @code-crusher this seems to be using a String literal still

**Comment by bklastaitis-branch** [comment]
@code-crusher commented back, I think there are still two things left `ServerResponse.requestId` and `Defines.Headers.RequestId("X-Branch-Request-Id")`.

However, as per Ahmed's comment, it seems like we should get in line between iOS and Android in how we log this. So if we happen to go the iOS route, then my feedback would be irrelevant. I'll leave it to you two decide which route is better but I left my 2 cents earlier.

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java
Comment (code-crusher) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterfaceUrlConnection.java:220]: Dunzo

**Comment by code-crusher** [comment]
@jdee doe the iOS prints in the above format I have put in the description? If not can you share one?

**Comment by code-crusher** [comment]
@bklastaitis-branch good to merge?

**Comment by bklastaitis-branch** [comment]
@code-crusher I was waiting on @jdee response regarding the format, but besides that, I think you still haven't addressed the comment about `ServerResponse.requestId`, let me know if doesn't make sense to you, or otherwise, if you disagree with that comment, happy to chat.

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java
Comment (code-crusher) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java]: > @code-crusher I was waiting on @jdee response regarding the format, but besides that, I think you still haven't addressed the comment about `ServerResponse.requestId`, let me know if doesn't make sense to you, or otherwise, if you disagree with that comment, happy to chat.

Used `Strings.emptyToNull(requestId)` -- just doesn't feel okay to add another step of processing, but it's fine with me.

**Comment by jdee** [comment]
Sorry I missed recent updates to this conversation. I've been updating the [tech spec](https://www.notion.so/brancheng/Request-ID-6c3c5c8525144a0cbae0e8da5567abc9) in Notion. Here's an example iOS log message.

```
2020-10-27 14:26:22.673178-0700 Branch-TestBed[61896:286008] [branch.io] BNCServerInterface.m(471) Debug: Server returned: [1325e434fa294d3bb7d461349118602d-2020102721] Status: 200; Data: {
    data = "{\"+clicked_branch_link\":false,\"+is_first_session\":false}";
    "device_fingerprint_id" = 848339731384240701;
    "identity_id" = 848339731459733646;
    link = "https://itrdp-jdee.branchbeta.link?%24identity_id=848339731459733646";
    "session_id" = 849748190046577385;
}.
```

**Comment by code-crusher** [comment]
@bklastaitis-branch looks good to merge?

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java
Comment (bklastaitis-branch) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java]: I think we are talking past each other @code-crusher 
> requestId should be a field within ServerResponse the same way it's a field in BranchResponse

I was saying that `ServerResponse` needs to have `requestId` passed in its constructor, as well as `BanchResponse`.. (I don't even know why we have `BanchResponse` in the first place, seems like we could just parse json directly to `ServerResponse`, but we can cross that bridge when we get there.)

**Comment by bklastaitis-branch** [comment]
@code-crusher https://github.com/BranchMetrics/android-branch-deep-linking-attribution/pull/869#discussion_r518247942

File: Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java
Comment (code-crusher) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java]: I don't see why that class needs to have this when BranchResponse is already passing it. What's the benefit of it that I'm missing?

Comment (bklastaitis-branch) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java]: > I still think that's the right thing to do because BranchResponse is just an intermediate class, if requestId was ever to be used in some other way besides logging, it would prolly would be referenced via ServerResponse instance which is further downstream.
...
> (I don't even know why we have BanchResponse in the first place, seems like we could just parse json directly to ServerResponse, but we can cross that bridge when we get there.)

Comment (code-crusher) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java]: But that's what I did earlier https://github.com/BranchMetrics/android-branch-deep-linking-attribution/pull/869/commits/f8fa8e3724a88f47b73205a1c1cedf3bdc1fa772 and mentioned the same: https://github.com/BranchMetrics/android-branch-deep-linking-attribution/pull/869#discussion_r513480024

Comment (code-crusher) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java]: Honestly, this is a too small change to think so much on this. I'll add the request id back into ServerResponse and update the PR

Comment (bklastaitis-branch) [reply Branch-SDK/src/main/java/io/branch/referral/network/BranchRemoteInterface.java]: Sorry for appearing picky, I think we just keep talking past each other.

I know you did that earlier and I agreed that it was a good idea, my [comment](https://github.com/BranchMetrics/android-branch-deep-linking-attribution/pull/869#discussion_r513480024) that you linked to was specifically about removing the setter and instead having the value passed in the constructor of `ServerResponse`.
>get rid of serverResponse.setRequestId_ and assign that value in the constructor, so it becomes a read-only variable, just like ServerResponse.statusCode_

**Comment by code-crusher** [comment]
@bklastaitis-branch added

**Review by bklastaitis-branch — APPROVED** [review]
@code-crusher thanks! FYI I added one more commit to make the print line identical to iOS (minus json formatting, which is present on iOS), hope that's fine. This is good to merge now but I'll leave that to you to make sure you saw my last message about the extra commit :)

**Comment by code-crusher** [comment]
Did you remove the request ID keyword too? I'm not really in favour of that since I don't see how any developer outside of Branch really know about it. But in any case, let's merge this and get it out.

**Comment by echo-branch** [comment]
I know this is a bit late but in the future I would strongly prefer to print raw server responses in verbose logging. Any effort to "pretty print" what we think is most useful is of limited value over raw logs and just complicates our code.

1. The raw response requires less code and less maintenance.  Every new header or field in responses will be logged automatically.
2. By the time verbose logging is enabled, the client is just going to copy paste the logs to us.  Client readability matters a lot less at that point.  While having more data is always useful to us.
3. Raw logs can be quickly tested via curl or postman.  This is of great value when debugging.  Pretty printed versions may omit data that is required to repro issues.

**Pull request merged**
